icicle/src/merkle-tree/mmcs.cu

```diff
@@ -0,0 +1,458 @@
+#include "hash/hash.cuh"
+#include "merkle-tree/merkle.cuh"
+#include "matrix/matrix.cuh"
+
+#include <algorithm>
+
+using matrix::Matrix;
+
+namespace merkle_tree {
+
+  template <typename L, typename D>
+  cudaError_t hash_leaves(
+    const Matrix<L>* leaves,
+    unsigned int number_of_inputs,
+    uint64_t number_of_rows,
+    D* states,
+    D* digests,
+    unsigned int digest_elements,
+    const SpongeHasher<L, D>& hasher,
+    const device_context::DeviceContext& ctx)
+  {
+    SpongeConfig sponge_config = default_sponge_config(ctx);
+    sponge_config.are_inputs_on_device = true;
+    sponge_config.are_outputs_on_device = true;
+    sponge_config.is_async = true;
+
+    uint64_t number_of_rows_padded = next_pow_of_two(number_of_rows);
+
+    std::cout << "Absorbing 2d" << std::endl;
+    CHK_IF_RETURN(hasher.absorb_2d(leaves, states, number_of_inputs, number_of_rows, ctx));
+    std::cout << "Squeezing" << std::endl;
+    CHK_IF_RETURN(hasher.squeeze_many(states, digests, number_of_rows, digest_elements, sponge_config));
+
+    if (number_of_rows_padded - number_of_rows) {
+      // Pad with default digests
+      cudaMemsetAsync(
+        (void*)digests, 0, (number_of_rows_padded - number_of_rows) * digest_elements * sizeof(D), ctx.stream);
+    }
+
+    return CHK_LAST();
+  }
+
+  template <typename L, typename D>
+  struct SubtreeParams {
+    unsigned int number_of_inputs;
+    unsigned int arity;
+    unsigned int digest_elements;
+    size_t number_of_rows;
+    size_t number_of_rows_padded;
+    size_t subtree_idx;
+    size_t number_of_subtrees;
+    uint64_t subtree_height;
+    size_t segment_size;
```

Comment:
what is segment? maybe add a comment

Reply:
segment is a memory segment that corresponds to a single layer of merkle tree digests